### PR TITLE
[release-0.5] Don't throw for tests which are explicitly marked broken

### LINKED ISSUE #21008: 0.5 `@test_skip` causes CI to fail
per https://discourse.julialang.org/t/how-to-use-test-skip-properly/2628

The following code will cause CI to fail, even though `@test_skip` should not contribute to CI success/failure:

```
using LightGraphs
using LightGraphs.LinAlg
using Base.Test

@testset "foo" begin
@test_skip 1 == 2
end
```
yields
```
julia> Pkg.test("LightGraphs")
INFO: Testing LightGraphs
Test Summary: | Broken  Total
  foo         |      1      1
ERROR: LoadError: Some tests did not pass: 0 passed, 0 failed, 0 errored, 1 broken.
```
causing a CI failure.

This is apparently fixed in 0.6, but I haven't been able to confirm that yet.

### PR DESCRIPTION
**Pull request by c42f**
Fixes #21008

Also grab the associated test case from #21086 (@tkelman - is that the right thing?  Do you want test cases backported along with minimal fixes?)

### CODE REVIEW

**Comment by tkelman** [comment]
do cherry-pick -x -e for the commit from the other PR and add a cross-reference to that PR in the commit message

**Comment by c42f** [comment]
Right, I added the cross refs to both patches.  https://github.com/JuliaLang/julia/pull/21087/commits/aa3907c5734b1a0292678c2b3c8ed77e8c18f40b is not a cherry pick, but a single line from a much larger refactoring.

**Review by StefanKarpinski — APPROVED** [review]

File: test/test.jl [hunk 1]
```diff
@@ -227,6 +227,18 @@ end
...
 
+# Issue #21008
+ts = try
+    @testset "@test_broken and @test_skip should not give an exception" begin
+        @test_broken false
+        @test_skip true
+        @test_skip false
+    end
+catch
+    nothing # Shouldn't get here
```
Comment (ggggggggg) [inline test/test.jl]: If you shouldn't get here, then shouldn't you make it cause a test failure if you do get there?

Comment (tkelman) [inline test/test.jl]: the result of the try catch is saved to a variable and its type is tested below

Comment (c42f) [inline test/test.jl]: Yes.  I can't remember exactly what I was thinking here, but it probably amounted to something like "`@test typeof(ts) == blah` is clearer than an `@test false` inside the catch".

Comment (ggggggggg) [inline test/test.jl]: Makes sense. Thanks for the explanation.

**Pull request merged**